README.md
# Translate Google
free google translate

## Change Log
- [changelog.md](https://github.com/shikar/NODE_GOOGLE_TRANSLATE/blob/master/CHANGELOG.md)


## Features

- Auto language detection
- Spelling correction
- Language correction
- Fast and reliable – it uses the same servers that [translate.google.com](https://translate.google.com) uses

## Install
```
npm install --save translate-google
```

## Usage

Perfect support object:

``` js
const translate = require('translate-google')
const tranObj = {
  a: 1,
  b: '1',
  c: "How are you?\nI'm nice.",
  d: [true, 'true', 'hi', { a: 'hello', b: ['world']}],
}

translate(tranObj, {to: 'zh-cn'}).then(res => {
    console.log(res)
}).catch(err => {
    console.error(err)
})

// => { a: 1, b: '1', c: "你好吗？\n我很好。", d: [true, 'true', '嗨', { a: '你好', b: ['世界']}] }
```

From automatic language detection to English:

``` js
const translate = require('translate-google')

translate('I speak Chinese', {to: 'zh-cn'}).then(res => {
    console.log(res)
}).catch(err => {
    console.error(err)
})
```

From English to Dutch with a typo:

``` js
translate('I speak Chinese!', {from: 'en', to: 'zh-cn'}).then(res => {
    console.log(res)
}).catch(err => {
    console.error(err)
})
```


translate for array or object:
``` js
translate({a: 'I speak Chinese!', b: ['hello', 'world']}, {from: 'en', to: 'zh-cn'}).then(res => {
    console.log(res)
}).catch(err => {
    console.error(err)
})
```

## API

### translate(text, options)

#### text

Type: `string`, `object`, `array`

The text to be translated

#### options

Type: `object`

##### from

Type: `string` Default: `auto`

The `text` language. Must be `auto` or one of the codes/names (not case sensitive) contained in [languages.js](https://github.com/shikar/NODE_GOOGLE_TRANSLATE/blob/master/languages.js)

##### to

Type: `string` Default: `en`

The language in which the text should be translated. Must be one of the codes/names (not case sensitive) contained in [languages.js](https://github.com/shikar/NODE_GOOGLE_TRANSLATE/blob/master/languages.js).

### Returns an `object`:

- `text` *(string, object, array)* – The translated text.

``` js
translate(['I speak Chinese\nHello world', 'hello'], {from: 'en', to: 'nl'}).then(res => {
    console.log(res);
    //=> ["我说中文\n你好世界","你好"]
}).catch(err => {
    console.error(err);
});
```

## License

MIT © [Shikar](qzh.shi@gmail.com)


CHANGELOG.md
# Change Log

## 1.4.2 [2020-07-28]
### Changed
- Changed url params processing method

## 1.4.1 [2020-07-27]
### Changed
- Change the submission module of post

## 1.4.0 [2020-07-16]
### Fixed
- Fixed Lots of translations bug

## 1.3.4 [2019-01-15]
### Added
- Added object translation exclusions
- Remove the duplicate part of the translation
### Fixed
- Fixed Object bug

## 1.3.4 [2019-01-12]
### Fixed
- Translate a large number of characters using POST method

## 1.3.3 [2019-01-10]
### Added
- Object mode translation: Excluding non-words that do not participate in translation

## 1.3.2 [2019-01-10]
### Added
- Object mode translation: Exclude numbers, URLs, keywords in translate

## 1.3.1 [2019-01-09]
### Fixed
- Fixed some bug

## 1.3.0 [2019-01-09]
### Changed
- Perfect support object


test.js
const test = require('ava')
const languages = require('./languages')
const translate = require('./index')

test('translate with string', async t => {
  try {
    const res = await translate('你好')
    t.is(res, 'Hello there')
  } catch (err) {
    t.log(err)
    t.fail()
  }
})

test('translate with array', async t => {
  try {
    const res = await translate(['你好', '你好'])
    t.is(res[0], 'Hello there')
    t.is(res[1], 'Hello there')
  } catch (err) {
    t.log(err)
    t.fail()
  }
})

test('translate with object', async t => {
  try {
    const res = await translate({a: '你好', b: '你好'})
    t.is(res.a, 'Hello there')
    t.is(res.b, 'Hello there')
  } catch (err) {
    t.log(err)
    t.fail()
  }
})

test('translate with full object', async t => {
  try {
    const obj = { a: 1, b: '1', c: "How are you?\nI'm nice.", d: [true, 'true', 'hi', { a: 'hello', b: ['world']}] }
    const res = await translate(obj, {from: 'en', to: 'zh-cn'})
    t.is(res.a, 1)
    t.is(res.b, '1')
    t.is(res.c, '你好吗？\n我很好。')
    t.is(res.d[0], true)
    t.is(res.d[1], 'true')
    t.is(res.d[2], '嗨')
    t.is(res.d[3].a, '你好')
    t.is(res.d[3].b[0], '世界')
  } catch (err) {
    t.log(err)
    t.fail()
  }
})

test('translate with array and object', async t => {
  try {
    const res = await translate({a: '你好', b: ['你好', '你好']})
    t.is(res.a, 'Hello there')
    t.is(res.b[0], 'Hello there')
    t.is(res.b[1], 'Hello there')
  } catch (err) {
    t.log(err)
    t.fail()
  }
})

test('translate with \\n', async t => {
  try {
    const res = await translate({a: '你\n好', b: ['你\n好', '你好']})
    t.is(res.a, 'you\nit is good')
    t.is(res.b[0], 'you\nit is good')
    t.is(res.b[1], 'Hello there')
  } catch (err) {
    t.log(err)
    t.fail()
  }
})

test('translate with except', async t => {
  try {
    const res = await translate({a: '你\n好', b: ['你\n好', '你好']}, {except:['a']})
    t.is(res.a, '你\n好')
    t.is(res.b[0], 'you\nit is good')
    t.is(res.b[1], 'Hello there')
  } catch (err) {
    t.log(err)
    t.fail()
  }
})

test('translate with option', async t => {
  try {
    const res = await translate('Hello', {from: 'en', to: 'zh-cn'})
    t.is(res, '你好')
  } catch (err) {
    t.log(err)
    t.fail()
  }
})

test('test a supported language – by code', t => {
  t.true(languages.isSupported('en'))
})

test('test an unsupported language – by code', t => {
  t.false(languages.isSupported('js'))
})


token.js
/*
* @Author: shikar
* @Date:   2017-02-05 13:50:16
* @Last Modified by:   shikar
* @Last Modified time: 2017-05-01 03:10:50
*/
'use strict'
const got = require('got')
const Configstore = require('configstore')

const config = new Configstore('google-translate')
const window = {
  TKK: config.get('TKK') || '0'
}

function sM(a) {
  var b
  if (null !== yr) {
    b = yr
  } else {
    b = wr(String.fromCharCode(84))
    var c = wr(String.fromCharCode(75))
    b = [b(), b()]
    b[1] = c()
    b = (yr = window[b.join(c())] || "") || ""
  }
  var d = wr(String.fromCharCode(116))
    , c = wr(String.fromCharCode(107))
    , d = [d(), d()]
  d[1] = c()
  c = "&" + d.join("") + "="
  d = b.split(".")
  b = Number(d[0]) || 0
  for (var e = [], f = 0, g = 0; g < a.length; g++) {
    var l = a.charCodeAt(g)
    128 > l ? e[f++] = l : (2048 > l ? e[f++] = l >> 6 | 192 : (55296 == (l & 64512) && g + 1 < a.length && 56320 == (a.charCodeAt(g + 1) & 64512) ? (l = 65536 + ((l & 1023) << 10) + (a.charCodeAt(++g) & 1023),
        e[f++] = l >> 18 | 240,
        e[f++] = l >> 12 & 63 | 128) : e[f++] = l >> 12 | 224,
        e[f++] = l >> 6 & 63 | 128),
        e[f++] = l & 63 | 128)
  }
  a = b
  for (f = 0; f < e.length; f++)
    a += e[f],
      a = xr(a, "+-a^+6")
  a = xr(a, "+-3^+b+-f")
  a ^= Number(d[1]) || 0
  0 > a && (a = (a & 2147483647) + 2147483648)
  a %= 1E6
  return c + (a.toString() + "." + (a ^ b))
}

var yr = null;
var wr = function(a) {
  return function() {
    return a
  }
}
  , xr = function(a, b) {
  for (var c = 0; c < b.length - 2; c += 3) {
    var d = b.charAt(c + 2)
      , d = "a" <= d ? d.charCodeAt(0) - 87 : Number(d)
      , d = "+" == b.charAt(c + 1) ? a >>> d : a << d
    a = "+" == b.charAt(c) ? a + d & 4294967295 : a ^ d
  }
  return a
}


function updateTKK(domain) {
  return new Promise( (resolve, reject) => {
    let now = Math.floor(Date.now() / 3600000)
    if (Number(window.TKK.split('.')[0]) === now) {
      resolve()
    } else {
      got(`https://${domain}`).then( res => {
        let code = res.body.match(/tkk:\'(.*?)\'/ig)
        let TKK = 0
        if (code) {
          TKK = code[0].match(/\d+\.\d+/)[0]
          if (typeof TKK !== 'undefined') {
            window.TKK = TKK
            config.set('TKK', TKK)
          }
        }

        resolve()
      }).catch( err => {
        let e = new Error()
        e.code = 'BAD_NETWORK'
        e.message = err.message
        reject(e)
      })
    }
  })
}

function get(text, domain = 'translate.google.cn') {
  return updateTKK(domain).then(() => {
    let tk = sM(text)
    tk = tk.replace('&tk=', '')
    return {name: 'tk', value: tk}
  }).catch( err => {
    throw err
  })
}

module.exports.get = get;


index.js
/*
* @Author: shikar
* @Date:   2017-02-05 15:28:31
* @Last Modified by:   shikar
* @Last Modified time: 2020-07-28 17:20:29
*/
'use strict'
const _ = require("lodash")
const isUrl = require("is-url")
const isNumber = require("num-or-not")
const isKeyword = require('is-keyword-js')
const got = require('got')
const userAgents = require("user-agents")
const token = require('./token')
const languages = require('./languages')

function checkSame(v, maps) {
  for (const idx in maps) {
    if (maps[idx].v === v) {
      return idx
    }
  }
  return -1
}

function enMap(obj, except=[], path='', map=[]) {
  if (_.isObject(obj) == true) {
    _.forEach(obj, (v, k) => {
      const furKeyStr = _.isNumber(k) ? `[${k}]` : ( path && '.' ) + k
      const curPath = path + furKeyStr
      if (_.isObject(v) == true) {
        enMap(v, except, curPath, map)
      } else {
        const exceptReg = except.length > 0 ? new RegExp(`(^|\\.)(${_.map(except, _.escapeRegExp).join('|')})(\\.|\\[|$)`, 'i') : false
        if (
          _.isString(v) &&
          !isNumber(v) &&
          !isUrl(v) &&
          !isKeyword(v) &&
          !/^(?!([a-z]+|\d+|[\?=\.\*\[\]~!@#\$%\^&\(\)_+`\/\-={}:";'<>,]+)$)[a-z\d\?=\.\*\[\]~!@#\$%\^&\(\)_+`\/\-={}:";'<>,]+$/i.test(v) &&
          (!exceptReg || !exceptReg.test(curPath))
        ) {
          const idx = checkSame(v, map)
          if (idx > -1) {
            map.splice(idx+1, 0, {
              p: curPath,
              v: v,
              i: map[idx].i,
              l: map[idx].l,
              s: true
            })
          } else {
            const lastMap = _.last(map)
            map.push({
              p: curPath,
              v: v,
              i: lastMap ? lastMap.i + lastMap.l : 0,
              l: v.split("\n").length,
              s: false
            })
          }
        }
      }
    })
  } else {
    map.push({
      p: '',
      v: obj,
      i: 0,
      l: obj.split("\n").length
    })
  }
  return map
}

function deMap(src, maps, dest) {
  if (_.isObject(src) == true) {
    src = _.clone(src)
    dest = dest.split("\n")
    for (const map of maps) {
      _.set(src, map.p, _.slice(dest, map.i, map.i+map.l).join("\n"))
    }
  } else {
    src = dest
  }
  return src
}

async function translate(input, opts = {}, domain='translate.google.cn') {
  const langs = [opts.from, opts.to]
  const except = opts.except || []
  for (const lang of langs) {
    if (lang && !languages.isSupported(lang)) {
      const e = new Error('The language \'' + lang + '\' is not supported')
      e.code = 400
      throw e
    }
  }

  opts.from = languages.getCode(opts.from || 'auto')
  opts.to = languages.getCode(opts.to || 'en')

  const strMap = enMap(input, except)
  const text = _.map(_.differenceBy(strMap, [{ s: true }], 's'), 'v').join("\n")
  const tokenRet = await token.get(text, domain)
  const url = `https://${domain}/translate_a/single`
  const searchParams = new URLSearchParams([
    ['client', 't'],
    ['sl', opts.from],
    ['tl', opts.to],
    ['hl', opts.to],
    ['dt', 'at'], ['dt', 'bd'], ['dt', 'ex'], ['dt', 'ld'], ['dt', 'md'], ['dt', 'qca'], ['dt', 'rw'], ['dt', 'rm'], ['dt', 'ss'], ['dt', 't'],
    ['ie', 'UTF-8'],
    ['oe', 'UTF-8'],
    ['otf', 1],
    ['ssel', 0],
    ['tsel', 0],
    ['kc', 7],
    ['q', text],
    [tokenRet.name, tokenRet.value]
  ])
  const opt = { responseType: 'json', headers: {'User-Agent': new userAgents({ deviceCategory: 'desktop' }).toString()} }
  if (searchParams.toString().length <= 1980) {
    opt.method = 'GET'
  } else {
    searchParams.delete('q')
    opt.method = 'POST'
    opt.form = { q: text }
  }
  opt.searchParams = searchParams
  try {
    const { body } = await got(url, opt)
    const retString = _.map(body[0], 0).join('')
    return deMap(input, strMap, retString)
  } catch (error) {
    let e = new Error(error.message)
    if (error.statusCode !== undefined && error.statusCode !== 200) {
      e.code = 'BAD_REQUEST'
    } else {
      e.code = 'BAD_NETWORK'
    }
    throw e
  }
}

module.exports = translate
module.exports.languages = languages
